specs/_features/eip7594/das-core.md

```diff
@@ -0,0 +1,285 @@
+    """
+    Return the recovered ``ExtendedMatrix``.
+
+    This helper demonstrates how to apply ``recover_polynomial``.
+    The data structure for storing cells is implementation-dependent.
+    """
+    extended_matrix = []
+    for blob_index in range(blob_count):
+        cell_ids = [cell_id for b_index, cell_id in cells_dict.keys() if b_index == blob_index]
+        cells = [cells_dict[(blob_index, cell_id)] for cell_id in cell_ids]
+        cells_bytes = [[bls_field_to_bytes(element) for element in cell] for cell in cells]
+
+        full_polynomial = recover_polynomial(cell_ids, cells_bytes)
+        cells_from_full_polynomial = [
+            full_polynomial[i * FIELD_ELEMENTS_PER_CELL:(i + 1) * FIELD_ELEMENTS_PER_CELL]
+            for i in range(CELLS_PER_BLOB)
+        ]
+        extended_matrix.extend(cells_from_full_polynomial)
+    return ExtendedMatrix(extended_matrix)
+```
+
+#### `get_data_column_sidecars`
+
+```python
+def get_data_column_sidecars(signed_block: SignedBeaconBlock,
+                             blobs: Sequence[Blob]) -> Sequence[DataColumnSidecar]:
+    signed_block_header = compute_signed_block_header(signed_block)
+    block = signed_block.message
+    kzg_commitments_inclusion_proof = compute_merkle_proof(
+        block.body,
+        get_generalized_index(BeaconBlockBody, 'blob_kzg_commitments'),
+    )
+    cells_and_proofs = [compute_cells_and_proofs(blob) for blob in blobs]
+    blob_count = len(blobs)
+    cells = [cells_and_proofs[i][0] for i in range(blob_count)]
+    proofs = [cells_and_proofs[i][1] for i in range(blob_count)]
+    sidecars = []
+    for column_index in range(NUMBER_OF_COLUMNS):
+        column = DataColumn([cells[row_index][column_index]
+                             for row_index in range(blob_count)])
+        kzg_proof_of_column = [proofs[row_index][column_index]
+                               for row_index in range(blob_count)]
+        sidecars.append(DataColumnSidecar(
+            index=column_index,
+            column=column,
+            kzg_commitments=block.body.blob_kzg_commitments,
+            kzg_proofs=kzg_proof_of_column,
+            signed_block_header=signed_block_header,
+            kzg_commitments_inclusion_proof=kzg_commitments_inclusion_proof,
+        ))
+    return sidecars
+```
+
+## Custody
+
+### Custody requirement
+
+Each node downloads and custodies a minimum of `CUSTODY_REQUIREMENT` subnets per slot. The particular subnets that the node is required to custody are selected pseudo-randomly (more on this below).
+
+A node *may* choose to custody and serve more than the minimum honesty requirement. Such a node explicitly advertises a number greater than `CUSTODY_REQUIREMENT` via the peer discovery mechanism -- for example, in their ENR (e.g. `custody_subnet_count: 4` if the node custodies `4` subnets each slot) -- up to a `DATA_COLUMN_SIDECAR_SUBNET_COUNT` (i.e. a super-full node).
```

Comment:
Implementation question - my apologies if it's off-topic for this PR - is there any value for a super-full node to custody the extended columns / rows (besides being able to serve data efficiently)? as it would always be able to reconstruct the full matrix without asking from its peers. 

If it stores just the raw blobs, it would only need to store 50% of the samples for 1D extended blobs, or just 25% of the samples of the 2D extended blobs:

| Max Blobs \ Data Stored | Storing raw blob only | 1D extended | 2D extended |
| --------- | -------- | ----------- | ----------- |
| 6         | 96 GB    | 192 GB      | 384 GB      |
| 32        | 512 GB   | 1 TB        | 2 TB        |
| 64        | 1TB      | 2 TB        | 4 TB        |
| 128       | 2TB      | 4 TB        | 8 TB        |

Reply:
> Implementation question - my apologies if it's off-topic for this PR - is there any value for a super-full node to custody the extended columns / rows (besides being able to serve data efficiently)? as it would always be able to reconstruct the full matrix without asking from its peers.

Reconstructing a column when you have only row encoding is quite expensive. Say you need column `j`. Then for every segment `i,j` in row `i`, you would need to read half of the row `i` to do the reconstruction. This basically means reading in the whole data amount to reconstruct a single row. 

So while you are always able to reconstruct, it can be costly, reading in and doing erasure coding on lots of data. I think it should be an implementation decision, but for the 1D extension I would keep the whole extended data for recent blocks, and only maybe discard half for old ones.

For the 2D extension, I would most probably store half of it, but in a more intelligent way than keeping the K columns or K rows. Basically what you want to do is to **keep the upper left corner** (original data in the systematic encoding), **and the lower right corner** ("double-parity" data) of the encoding. This allows you to efficiently restore any row, column or individual segment, while keeping only half of the data. Again, for old data that most probably won't be queried, you might want to fall back to keeping only the original (or some other K half-rows), while paying the price of a full restore when queried.

We might even go further, and instead of keeping the upper left corner and lower right corner, select a set `A` of K row IDs and a set `B` of K column IDs. Then you store chunk `i,j` iff
$i \in A \land j \in B \lor i \notin A \land j \notin B$
This allows the same fast reconstruction, but randomises who stores what. This can be done with local randomness to improve resilience, or with peerID based randomness to reduce load of queries from good behaving nodes who can query nodes where they know the node can serve cheap without reconstruction.

Of course, a super-full node would have to balance the above with its own needs to access the original data. If you actually want to read the original data frequently for your own use, it would be stupid not to store the original part (upper left corner) of the systematic encoding.

Reply:
> Implementation question - my apologies if it's off-topic for this PR - is there any value for a super-full node to custody the extended columns / rows (besides being able to serve data efficiently)? as it would always be able to reconstruct the full matrix without asking from its peers.
> 
> If it stores just the raw blobs, it would only need to store 50% of the samples for 1D extended blobs, or just 25% of the samples of the 2D extended blobs:
> Max Blobs \ Data Stored 	Storing raw blob only 	1D extended 	2D extended
> 6 	96 GB 	192 GB 	384 GB
> 32 	512 GB 	1 TB 	2 TB
> 64 	1TB 	2 TB 	4 TB
> 128 	2TB 	4 TB 	8 TB

All these are off by a factor of 2, assuming that target blobs are 1/2 of max blobs. (The EIP-1559 style pricing mechanism makes it impossible to significantly exceed the target for such a long period by increasing the price exponentially)